File: frontend/src/lib/answerSections.ts
/**
 * 把综述正文按 `core/ask.py` `SYN_SYS` 约定的四个标签切成分节，供答案页分模块渲染。
 *
 * 真实语料存在三种标签形态：`**标签**` 独占一行、`**标签** — 正文`、`**标签**:` +
 * 行尾硬换行；模型偶尔改用 `## 标签` 标题。切分只识别标签本身，模块的标题文字与
 * 图标由渲染层的固定元数据决定（见 `SectionHeading`），不复用模型写的字面量。
 */
export type SectionKind =
  'conclusion' | 'evidence' | 'picos' | 'caveats' | 'other'

export type AnswerSection = { kind: SectionKind; markdown: string }

/** `**标签**`（可带 `#` 前缀）后可选 破折号/冒号 分隔符，组 2 为同行余文。 */
const BOLD_LABEL =
  /^\s*(?:#{1,6}\s+)?\*\*([^*\n]+?)\*\*\s*(?:[—–\-:：]\s*)?(.*)$/
/** `## 标签` 形态，无同行余文。 */
const HEADING_LABEL = /^\s*#{1,6}\s+([^\n]+?)\s*$/

/**
 * 标签文字 → 分节类型。全部按「前缀」匹配：真实标签都以关键词开头
 * （`结论 / Bottom line`、`证据 / Evidence`、`PICOS 证据表 / PICOS table`、
 * `局限 / Caveats`）。用包含匹配会把 `# Q: …的证据强度` 这类问题标题误判成分节头。
 * 判定顺序不可调整：picos 先于 evidence。
 */
export function sectionKindOf(label: string): SectionKind | null {
  const l = label.trim()
  const lower = l.toLowerCase()
  if (l.startsWith('结论') || lower.startsWith('bottom line'))
    return 'conclusion'
  if (lower.startsWith('picos')) return 'picos'
  if (l.startsWith('证据') || lower.startsWith('evidence')) return 'evidence'
  if (
    l.startsWith('局限') ||
    lower.startsWith('caveat') ||
    lower.startsWith('limitation')
  )
    return 'caveats'
  return null
}

function sectionHead(line: string): { kind: SectionKind; rest: string } | null {
  const bold = BOLD_LABEL.exec(line)
  if (bold) {
    const kind = sectionKindOf(bold[1] ?? '')
    if (kind) return { kind, rest: bold[2] ?? '' }
  }
  const heading = HEADING_LABEL.exec(line)
  if (heading) {
    const kind = sectionKindOf(heading[1] ?? '')
    if (kind) return { kind, rest: '' }
  }
  return null
}

/**
 * 按分节头切分正文，保持文档顺序。空白节丢弃；同一个已知 kind 重复出现时合并到
 * 首次出现的节（`\n\n` 连接），`other` 可出现多次（前言或未识别标签之间的内容）。
 */
export function splitAnswerSections(markdown: string): AnswerSection[] {
  const sections: AnswerSection[] = []
  let kind: SectionKind = 'other'
  let buffer: string[] = []
  const flush = () => {
    const md = buffer.join('\n').trim()
    buffer = []
    if (!md) return
    const existing =
      kind === 'other' ? undefined : sections.find((s) => s.kind === kind)
    if (existing) existing.markdown += '\n\n' + md
    else sections.push({ kind, markdown: md })
  }
  for (const line of markdown.replace(/\r\n/g, '\n').split('\n')) {
    const head = sectionHead(line)
    if (!head) {
      buffer.push(line)
      continue
    }
    flush()
    kind = head.kind
    if (head.rest) buffer.push(head.rest)
  }
  flush()
  return sections
}

/** 正文里是否出现过已识别的分节头；否则整段走单块渲染回退。 */
export function hasKnownSections(sections: AnswerSection[]): boolean {
  return sections.some((s) => s.kind !== 'other')
}

File: flutter/lib/core/logic/answer_sections.dart
// 答案正文分节：把模型输出的粗体标签行（`**结论 / Bottom line**` 等）切成模块，
// 供答案页按「结论卡 / 证据 / PICOS / 局限」分别渲染。三端算法逐字一致。

/// 一个模块的类型；`other` 是前言或未识别标签之间的内容。
enum AnswerSectionKind { conclusion, evidence, picos, caveats, other }

/// 一节正文：[kind] 决定渲染容器，[markdown] 是去首尾空白后的原始 Markdown。
class AnswerSection {
  const AnswerSection(this.kind, this.markdown);

  final AnswerSectionKind kind;
  final String markdown;

  @override
  bool operator ==(Object other) =>
      other is AnswerSection &&
      other.kind == kind &&
      other.markdown == markdown;

  @override
  int get hashCode => Object.hash(kind, markdown);

  @override
  String toString() => 'AnswerSection(${kind.name}, $markdown)';
}

/// 粗体标签行：`**标签**`、`**标签** — 正文`、`**标签**: 正文`，允许前置 `#` 标题标记。
final _boldLabelRegex = RegExp(
  r'^\s*(?:#{1,6}\s+)?\*\*([^*\n]+?)\*\*\s*(?:[—–\-:：]\s*)?(.*)$',
);

/// 纯标题行：`## 结论`。
final _headingLabelRegex = RegExp(r'^\s*#{1,6}\s+([^\n]+?)\s*$');

/// 标签文字 → 模块类型；不是已知模块返回 null。
///
/// 四类一律「前缀」匹配：问题标题里出现关键词（如 `# Q: … 的证据强度`）不能被
/// 当成分节头。PICOS 必须先判：标签「PICOS 证据表 / PICOS table」也含「证据」。
AnswerSectionKind? answerSectionKind(String label) {
  final l = label.trim();
  final lower = l.toLowerCase();
  if (l.startsWith('结论') || lower.startsWith('bottom line')) {
    return AnswerSectionKind.conclusion;
  }
  if (lower.startsWith('picos')) return AnswerSectionKind.picos;
  if (l.startsWith('证据') || lower.startsWith('evidence')) {
    return AnswerSectionKind.evidence;
  }
  if (l.startsWith('局限') ||
      lower.startsWith('caveat') ||
      lower.startsWith('limitation')) {
    return AnswerSectionKind.caveats;
  }
  return null;
}

/// 按分节头切分正文。无分节头时整段作为一个 `other` 节；空输入返回 `[]`。
List<AnswerSection> splitAnswerSections(String markdown) {
  final out = <AnswerSection>[];
  final buffer = <String>[];
  var kind = AnswerSectionKind.other;

  void flush() {
    final text = buffer.join('\n').trim();
    buffer.clear();
    if (text.isEmpty) return;
    if (kind != AnswerSectionKind.other) {
      // 同类模块重复出现（模型分两段写证据）→ 合并进已有节，不新增。
      final index = out.indexWhere((section) => section.kind == kind);
      if (index >= 0) {
        out[index] = AnswerSection(kind, '${out[index].markdown}\n\n$text');
        return;
      }
    }
    out.add(AnswerSection(kind, text));
  }

  for (final line in markdown.replaceAll('\r\n', '\n').split('\n')) {
    var label = '';
    var rest = '';
    final bold = _boldLabelRegex.firstMatch(line);
    if (bold != null) {
      label = bold.group(1) ?? '';
      rest = bold.group(2) ?? '';
    } else {
      final heading = _headingLabelRegex.firstMatch(line);
      if (heading != null) label = heading.group(1) ?? '';
    }
    final next = label.isEmpty ? null : answerSectionKind(label);
    if (next == null) {
      buffer.add(line);
      continue;
    }
    flush();
    kind = next;
    if (rest.isNotEmpty) buffer.add(rest);
  }
  flush();
  return out;
}

/// 是否识别出了任何已知模块；否则整段按单块渲染（旧答案、模型没按格式写）。
bool hasKnownSections(List<AnswerSection> sections) =>
    sections.any((section) => section.kind != AnswerSectionKind.other);
